### DISCUSSION TITLE
Can mitmdump be used to modify QUIC packets

### DISCUSSION THREAD

**Discussion by ddworkin** [mitmproxy/mitmproxy#6448] [Q&A]
We are working on  an app which makes a Quic connection to a server.  It is _not_ doing HTTP3 - we are using Quic as a transport, creating multiple streams over a single quic connection.  

I want to simulate loss on one more more individual stream connections (so that we can test how it behaves and make sure everything is working well), so I would like to modify the data being sent from the client - if I can access the data, I can figure out which stream the data belongs to and randomly drop the packets to simulate loss.   I tried to see if I can do so with a python script but it seems that the quic support is not really there yet.

I was able to have these functions called:
_def quic_start_client(data: mitmproxy.proxy.layers.quic.QuicTlsData):
def quic_start_server(data: mitmproxy.proxy.layers.quic.QuicTlsData):_

but I believe in order to modify the data (for example to simulate loss) I would need to implement this:
_def modify(data: bytes) -> bytes | Iterable[bytes]:_
But that doesn't seem to be getting called for Quic.

the command I'm running is:
**mitmdump -p 8000 --mode reverse:quic://127.0.0.1:4567 -s myscript.py**

Is this just not supported (yet), or is there a way for me to do what I'm trying?
Thanks

**Comment by mhils** [comment]
Hi there! 

Generally speaking mitmproxy can work with (and modify) raw QUIC streams. mitmproxy will handle individual raw QUIC streams as TCP flows internally[^1]. Here's an example with a simple addon and the corresponding mitmproxy/curl invocations:

```
mitmdump -s example.py --mode reverse:quic://mitmproxy.org
curl --http3 https://mitmproxy.org/robots.txt --connect-to mitmproxy.org:443:localhost:8080
```
```python
def tcp_message(flow):
    flow.messages[-1].content = flow.messages[-1].content.replace(
        b"Disallow:",
        b"Allow:   ",
    )
```

That being said, this really isn't a suitable mechanism to simulate packet loss. Generally speaking, mitmproxy doesn't lend itself to simulate that as we are terminating TCP/TLS/QUIC/... at the application layer. You can use the mechanism above to permanently modify the contents of a stream, but this won't trigger retransmissions.


I suppose you could use `tcp_message` to identify transmissions of interest and then selectively drop the next UDP packet by monkeypatching [here](https://github.com/mitmproxy/mitmproxy/blob/10.1.3/mitmproxy/proxy/server.py#L395), but this is getting _very_ hacky very quickly.

[^1]: It obviously isn't TCP and the naming here is a bit confusing; this is because we wanted to reuse our existing "raw stream of data" mechanisms and not break backward comp.

**Comment by ddworkin** [comment]
Thanks for the response.   I thought it would trigger retransmission - the packets/frames are sequenced, so if I intercept in either direction and identify a stream (which would require me to parse the data so that I can identify a 1-RTT packet with a STREAM frame and find the stream id) then when it gets forwarded to the destination it would not be complete.

(Since mitmproxy understands quic - it displays stream number for example - I was hoping to not even need to parse the data so much.  While I’m running my tests, mitmproxy displays:
127.0.0.1:60204 -> quic stream 1 -> mitmproxy -> quic stream 1 -> 127.0.0.1:4567
So it clearly understands the quic data)

It seems that flow.messages[-1].content gives me the payload but not the raw data, so I can’t tell (for example) which stream it is.

I tried adding the tcp_message as you suggest, and just tried dropping the first two bytes
    flow.messages[-1].content = flow.messages[-1].content[2:]
but as you say, it didn’t trigger anything (the quic header has the frame length:
STREAM Frame {
Type (i) = 0x08..0x0f,
Stream ID (i),
[Offset (i)],
[Length (i)],
Stream Data (..),
}
But what I’m afraid is happening is, the length is optional – if the length field is not there, the stream data extends to the end of the packet.  So in my case, it’s probably not there (I don’t know for sure since I can’t see the quic headers), and all I’m doing is truncating my payload

I ran into a few other problems as well –

When playing around with mitmproxy, I was able to get the aioquic stack to crash in a few different ways so it might not be robust enough (at this point) for me to use even if I can come up with a nice way to simulate loss.

Case 1 – my quic client is sending a small payload (around 64 bytes), but I’m doing it in a loop.  If I only loop 10 or 20 times it’s ok, but if I loop 1000 times, I get this:

/opt/homebrew/Cellar/mitmproxy/10.0.0/libexec/lib/python3.12/site-packages/aioquic/_crypto.py:366<http://crypto.py:366>: UserWarning: ffi.buffer(cdata, bytes): creating a buffer of 9185 bytes over a cdata that owns only 1500 bytes.  This will crash if you access the extra memory
  self._binding.ffi.buffer(self._buffer, packet_len)[:] = packet

Case 2 – my quic client is sending a single large payload – in this case, it was around 2500 bytes.  I get the same error as Case 1
/opt/homebrew/Cellar/mitmproxy/10.0.0/libexec/lib/python3.12/site-packages/aioquic/_crypto.py:366<http://crypto.py:366>: UserWarning: ffi.buffer(cdata, bytes): creating a buffer of 2538 bytes over a cdata that owns only 1500 bytes.  This will crash if you access the extra memory
  self._binding.ffi.buffer(self._buffer, packet_len)[:] = packet
Segmentation fault: 11

(Even if I can’t simulate loss, mitmproxy allowed me to show the quic traffic in wireshark which is still helpful for debugging and troubleshooting, but I can’t use it due to the crashes)

Thanks again,

David


From: Maximilian Hils ***@***.***>
Date: Saturday, November 4, 2023 at 7:58 AM
To: mitmproxy/mitmproxy ***@***.***>
Cc: ddworkin ***@***.***>, Author ***@***.***>
Subject: Re: [mitmproxy/mitmproxy] Can mitmdump be used to modify QUIC packets (Discussion #6448)

Hi there!

Generally speaking mitmproxy can work with (and modify) raw QUIC streams. mitmproxy will handle individual raw QUIC streams as TCP flows internally1. Here's an example with a simple addon and the corresponding mitmproxy/curl invocations:

mitmdump -s example.py --mode reverse:quic://mitmproxy.org

curl --http3 https://mitmproxy.org/robots.txt --connect-to mitmproxy.org:443:localhost:8080

def tcp_message(flow):

    flow.messages[-1].content = flow.messages[-1].content.replace(

        b"Disallow:",

        b"Allow:   ",

    )

That being said, this really isn't a suitable mechanism to simulate packet loss. Generally speaking, mitmproxy doesn't lend itself to simulate that as we are terminating TCP/TLS/QUIC/... at the application layer. You can use the mechanism above to permanently modify the contents of a stream, but this won't trigger retransmissions.

I suppose you could use tcp_message to identify transmissions of interest and then selectively drop the next UDP packet by monkeypatching here<https://github.com/mitmproxy/mitmproxy/blob/10.1.3/mitmproxy/proxy/server.py#L395>, but this is getting very hacky very quickly.

Footnotes

  1.  It obviously isn't TCP and the naming here is a bit confusing; this is because we wanted to reuse our existing "raw stream of data" mechanisms and not break backward comp. ↩

**Reply by mhils** [reply]
> I thought it would trigger retransmission

You're not operating on packets here, you're operating on _stream contents_. Packet numbers/lengths are adjusted transparently. As I said, it's not really meant for your use case.

>When playing around with mitmproxy, I was able to get the aioquic stack to crash in a few different ways

/cc @meitinger - this seems to be in _crypto.py, which is not stock aioquic? 🙈

**Reply by meitinger** [reply]
Sorry for the late reply. This should be fixed in https://github.com/meitinger/aioquic_mitmproxy/commit/ad23a04ba7cba11100f7be33b4cbbd75ca3d5217.

**Reply by ddworkin** [reply]
@meitinger great!  How can I use the fixes? (I currently have mitmproxy installed via _brew_)

**Comment by ddworkin** [comment]
Also I notice my quic connection is timing out…I’m getting my data through but then shortly after, it times out. I have to see if there’s any trick to prevent it (I don’t think I’m doing anything to cause that in n my test client/server)On Nov 4, 2023, at 12:14 PM, Maximilian Hils ***@***.***> wrote:﻿

I thought it would trigger retransmission

You're not operating on packets here, you're operating on stream contents. Packet numbers/lengths are adjusted transparently. As I said, it's not really meant for your use case.

When playing around with mitmproxy, I was able to get the aioquic stack to crash in a few different ways

/cc @meitinger - this seems to be in _crypto.py, which is not stock aioquic? 🙈
